### ISSUE TITLE
regarding the usage of gradients and self.net.gradients_nodes

### ISSUE THREAD

**Issue by surfreta** [jakeret/tf_unet#3]
Hi Joel, 

The previous thread has become so long. I just closed it. Thanks for your help as always.

I am re-running the code with batch size =2, which was reported by some blogs to get better result. At the same time, I am trying to normalize the image as you suggested.

Besides, I am still feeling confused with the line of code to perform the optimization. My understanding for your including `gradients` and `self.net.gradients_nodes` is for the debugging usages. 

` _, loss, lr, gradients = sess.run((self.optimizer, self.net.cost, self.learning_rate_node, self.net.gradients_node), 
                                                      feed_dict={self.net.x: batch_x,
                                                                 self.net.y: util.crop_to_shape(batch_y, pred_shape),
                                                                 self.net.keep_prob: dropout})`

In other words, if I am only trying to run the training process, I can remove these two items and re-write your code as. Is my understanding right? In specific, I think I am not very clear about the usage of `gradients` here.  Thanks.

` _, loss, lr= sess.run((self.optimizer, self.net.cost, self.learning_rate_node), 
                                                      feed_dict={self.net.x: batch_x,
                                                                 self.net.y: util.crop_to_shape(batch_y, pred_shape),
                                                                 self.net.keep_prob: dropout})`

**Comment by jakeret** [comment]
> My understanding for your including gradients and self.net.gradients_nodes is for the debugging usages.

If `Unet.summaries` is set to true you get some histograms in the Tensorboard displaying the gradient values for each epoch. If the values remain very large for instance it might indicate that the training is not working.

> In other words, if I am only trying to run the training process, I can remove these two items and re-write your code as.

Yes in principle, but I still don't understand why you want to make changes in the `Trainer` class

**Comment by surfreta** [comment]
Hi Joel,

No, I do not plan to change the trainer class. The reason that I am asking this question is because, I saw some other tensorflow implementations, and found that most of them normally do not include the gradient part explicitly in the optimization( or back-propagation) part as you did. I am just curious the reason that you want to do this way, or any trick(or special considerations) for this? In accordance with your explanation, the purpose is more for debugging and performance tracking purposes.

**Comment by jakeret** [comment]
Out of curiosity I downloaded the kaggle ultrasound dataset. To start the training I reused most of the `script/rfi_launcher.py" and `script/radio_util.py` code. I more or less just replaced the loading of the data with a call to the PIL library: 

`data = np.array(Image.open(self.data_files[self.file_idx]), dtype=np.float32)`

I don't get black predictions and the learning rate is decreasing as expected

**Comment by surfreta** [comment]
Hi Joel,
Thanks for trying this data set as well. I really want to fully understand your code and get it worked because I really like your code, which is very well designed.

Do you use the same parameter, such as three layers, feature_root=16, Have you normalized the image?
I don’t change the optimization step. The only thing I change is the code for loading the image files. I am quite confused why my ones does not work.

(a quick update, i just got the figure after finishing 12 epochs, looks like the prediction one is not full black. But the learning rate does not change as well. )
![epoch_12](https://cloud.githubusercontent.com/assets/23064201/20197208/dd6b01e0-a763-11e6-961c-705b07196f1e.jpg)


In the demo problem, generally, you randomly generate training data set on the fly within each epoch. For kaggle case, I need to iterate over the training set consisting of 5635 pairs of images and mask images

This is what I did during training step, I use `batch_x, batch_y = data_provider(ij, self.batch_size)  `  to replace `batch_x, batch_y = data_provider(self.batch_size)`.  Here ij represents the starting index because I need to iterate over the training set. 

The following is [new_get_image_gen](url),  which is similar to `get_image_gen`. The only difference is that I introduce one input parameter to control the starting point for each iteration

`def new_get_image_gen(raw_image, mask_image):    
    def new_create_batch(starting_point, n_image):
    #starting_point to keep track the iteraion process over the input
           batch_x = raw_image[starting_point:starting_point+n_image,:,:,:]
           batch_y = mask_image[starting_point:starting_point+n_image,:,:,:]
    
           return batch_x,batch_y
        
    new_create_batch.channels = 1
    new_create_batch.n_class = 2
    return new_create_batch`

This is how I launch the training process

`image_gen.create_train_data(raw_data_path,train_data_path)  # only the first time needs to include this operation
    imgs_train = np.load(os.path.join(train_data_path,"imgs_train.npy"))
    imgs_mask_train = np.load(os.path.join(train_data_path,"imgs_mask_train.npy"))   
    
    generator = image_gen.new_get_image_gen(imgs_train,imgs_mask_train)
    net = unet.Unet(channels=generator.channels, n_class=generator.n_class, layers=3, features_root=16)
    trainer = unet.Trainer(net, optimizer="momentum", opt_kwargs=dict(momentum=0.2))
    path = trainer.train(generator, "/data/unet_trained", training_iters=2816, epochs=100, display_step=2)
    `
This is  `create_train_data`. Generally, it will return two arrays: one is `imgs_train of shape (5635,420,580,1) `and `imgs_mask_train of shape(5635,420,580,2)`.

`def create_train_data(raw_data_path,train_data_path):   
    image_rows = 420
    image_cols = 580 
    images = os.listdir(raw_data_path)
    total = int(len(images) / 2)

    imgs = np.ndarray((total,image_rows, image_cols,1), dtype=np.uint8)
    imgs_mask = np.ndarray((total,image_rows, image_cols,2), dtype=np.uint8)

    i = 0
    print('-'*30)
    print('Creating training images...')
    print('-'*30)
    for image_name in images:
        if 'mask' in image_name:
            continue
        image_mask_name = image_name.split('.')[0] + '_mask.tif'
        img = io.imread(os.path.join(raw_data_path, image_name))
        img_mask = io.imread(os.path.join(raw_data_path, image_mask_name))
        img_mask = img_mask//255
        img_mask_background = 1-img_mask
        

        img = np.array([img])
        img_mask = np.array([img_mask])
        img_mask_background = np.array([img_mask_background])

        imgs[i] = img.reshape(image_rows,image_cols,1)
        imgs_mask[i] = np.concatenate((img_mask.reshape(image_rows,image_cols,1),img_mask_background.reshape(image_rows,image_cols,1)),axis=-1)

        if i % 100 == 0:
            print('Done: {0}/{1} images'.format(i, total))
        i += 1
    print('Loading done.')
    
    np.save(os.path.join(train_data_path,"imgs_train.npy"),imgs)
    np.save(os.path.join(train_data_path,"imgs_mask_train.npy"),imgs_mask)
    
    print('Saving to .npy files done.')`

**Comment by jakeret** [comment]
Here is my implementation: https://gist.github.com/jakeret/cd1e8b51009b995c38b44101c38ad831

To create the instance I use something like this

```python
from scripts.ultrasound_util import DataProvider

data_provider = DataProvider(data_files=glob.glob(data_root+"/[0-9]?_[0-9]?.tif"),
                             mask_files=glob.glob(data_root+"/[0-9]?_[0-9]?_mask.tif"),)
```
Just like in the rfi example

**Comment by surfreta** [comment]
Hi Joel,

Thanks.
